app/controllers/tags_controller.rb

```diff
@@ -117,6 +117,35 @@ def update
     end
   end
 
+  def upload
+    guardian.ensure_can_admin_tags!
+
+    file = params[:file] || params[:files].first
+
+    hijack do
+      begin
+        Tag.transaction do
+          CSV.foreach(file.tempfile) do |row|
+            raise Discourse::InvalidParameters.new(I18n.t("tags.upload_row_too_long")) if row.length > 2
+
+            tag_name = DiscourseTagging.clean_tag(row[0])
+            tag_group_name = row[1] || nil
+
+            tag = Tag.find_by_name(tag_name) || Tag.create!(name: tag_name)
+
+            if tag_group_name
+              tag_group = TagGroup.find_by(name: tag_group_name) || TagGroup.create!(name: tag_group_name)
+              tag.tag_groups << tag_group unless tag.tag_groups.include?(tag_group)
+            end
+          end
+        end
+        render json: success_json
+      rescue => e
```

Comment:
Do we need which errors we want to rescue here? If so we should only rescue those errors and let the rest blow up. Otherwise, `rescue` will rescue things like typos as well.

Reply:
Thanks, I have explicitly rescued InvalidParameters (the generic `rescue_from` in `application_controller` doesn't apply inside a hijack block)